js/style/style_layer/fill_style_layer.js

```diff
@@ -7,6 +7,48 @@ function FillStyleLayer() {
     StyleLayer.apply(this, arguments);
 }
 
-module.exports = FillStyleLayer;
+FillStyleLayer.prototype = util.inherit(StyleLayer, {
+
+    getPaintValue: function(name, globalProperties, featureProperties) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.getPaintValue.call(this, 'fill-color', globalProperties, featureProperties);
+        } else {
+            return StyleLayer.prototype.getPaintValue.call(this, name, globalProperties, featureProperties);
+        }
+    },
+
+    getPaintValueStopZoomLevels: function(name) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.getPaintValueStopZoomLevels.call(this, 'fill-color');
+        } else {
+            return StyleLayer.prototype.getPaintValueStopZoomLevels.call(this, arguments);
+        }
+    },
+
+    getPaintInterpolationT: function(name, zoom) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.getPaintInterpolationT.call(this, 'fill-color', zoom);
+        } else {
+            return StyleLayer.prototype.getPaintInterpolationT.call(this, name, zoom);
+        }
+    },
 
-FillStyleLayer.prototype = util.inherit(StyleLayer, {});
+    isPaintValueFeatureConstant: function(name) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.isPaintValueFeatureConstant.call(this, 'fill-color');
+        } else {
+            return StyleLayer.prototype.isPaintValueFeatureConstant.call(this, name);
+        }
+    },
+
+    isPaintValueZoomConstant: function(name) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.isPaintValueZoomConstant.call(this, 'fill-color');
+        } else {
+            return StyleLayer.prototype.isPaintValueZoomConstant.call(this, name);
+        }
+    }
+
+});
```

Comment:
I'm not sold on that design 

 - Similar branching logic would still need to be written, but it'd be in `StyleLayer` rather than `FillLayer`
 - I can't think of many (any?) other style properties that have fallback behaviour like this

----

My preferred solution would be to factor out all the per-paint-property methods in `StyleLayer` into a `StyleValue` class (that might replace `StyleDeclaration` and/or `StyleTransition`)

```js
type StyleValue<T> {
  isZoomConstant: boolean;
  isFeatureConstant: boolean;
  calculate(zoom:number, featureProperties: {}) => T;
  getInterpolationT(zoom: number) =>;
  ...
}
```

Then, `FillStyleLayer`'s implementation would only have one overriden method: 

```js
getStyleValue: function(name) {
  if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
    return StyleLayer.prototype.getStyleValue.call(this, 'fill-color');
  } else {
    return StyleLayer.prototype.getStyleValue.call(this, name);
  }
}
```

----

All that said, I'd be happy to 🚢 with the code in this PR for now. It may be verbose and non-DRY but it is clear, robust, and well-contained.